TwinePlatform/twine-monolith
# 👀 implement monitoring of user sessions

fixes #255 

### Changes
- Add step to release phase to configure redis
- Add `user_session_record` table and `UserSessionRecords` collection
- Update `user_session_record` when users:
  - Login
  - Make authenticated requests to the API
  - Logout
  - Session expire
- Add `ioredis` dependency
- Add listener when redis session keys expire (and update `user_session_record`)
- Add session monitoring assertions to `auth_flow` test
- Add `Promises.silent` util which ignores rejections for auxiliary actions
- Remove calls to insert into `user_account_active_day` and `login_event` -- this is can be calculated from `user_session_record`

### Testing Requirements
  - Login (then check `user_session_record` is created)
  - Navigate to various pages (then check `user_session_record.referrers` is updated)
  - Wait for the session to expire (then check `user_session_record.session_end_type = 'expired'`)
  - Do the above flow again but log out instead
- [x] Visitor app
  - As for dashboard
  - As for dashboard

### Release Requirements
To be confirmed

### Manual Deployment
- API
